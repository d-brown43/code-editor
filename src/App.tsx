import * as React from 'react';
import CodeEditor from './CodeEditor';
import styles from './App.module.scss';
import defaultProgram from "./defaultProgram";
import Controls from "./Controls";
import {compileForErrors, prepareWindow, run} from "./compilation";
import ConsoleLog from './ConsoleLog';
import useLogs from "./useLogs";
import useGlobalReplacements from "./useGlobalReplacements";

type Programs = {
    [key: string]: string;
}

type ProgramErrors = {
    [key: string]: ProgramError;
}

const App = () => {
    const [activeProgram, setActiveProgram] = React.useState<string>('index');
    const [logs, addLogMessage] = useLogs();
    const [programs, setPrograms] = React.useState<Programs>({
        index: defaultProgram()
    });
    const [programErrors, setProgramErrors] = React.useState<ProgramErrors>({});
    const {globalReplacements, cleanup} = useGlobalReplacements(addLogMessage);

    React.useEffect(() => {
        prepareWindow(globalReplacements);
    }, [globalReplacements]);

    const setProgramGenerator = (key: string) => (program: string) => {
        setPrograms((prevPrograms) => ({
            ...prevPrograms,
            [key]: program
        }))
    };

    React.useEffect(() => {
        setProgramErrors(Object.entries(programs).reduce((result, [programKey, program]) => ({
            ...result,
            [programKey]: compileForErrors(program)
        }), {}));
    }, [programs]);

    return (
        <div className={styles.container}>
            <Controls
                run={() => {
                    try {
                        cleanup();
                        run(programs.index, globalReplacements);
                    } catch (e) {
                        addLogMessage('error')(e.message.replace(/__codeEditor__\./g, ''));
                    }
                }}
                stop={cleanup}
            />
            {
                Object.entries(programs).map(([filename, program]) => (
                    <CodeEditor
                        key={filename}
                        program={program}
                        setProgram={setProgramGenerator(filename)}
                        programError={programErrors[activeProgram]}
                    />
                ))
            }
            <ConsoleLog logMessages={logs} />
        </div>
    );
};

export default App;
