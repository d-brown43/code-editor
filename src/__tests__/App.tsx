import * as ReactDOM from "react-dom";
import App from "../App";
import * as React from "react";
import {shallow} from "enzyme";

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
});

it('matches snapshot', () => {
    const app = shallow(<App />);
    expect(app).toMatchSnapshot();
});
