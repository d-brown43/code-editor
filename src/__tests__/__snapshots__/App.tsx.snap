// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="container"
>
  <Controls
    run={[Function]}
    stop={[Function]}
  />
  <CodeEditor
    key="index"
    program="console.log('Hello World');


"
    setProgram={[Function]}
  />
  <ConsoleLog
    logMessages={Array []}
  />
</div>
`;
